support/k8s/k8s-workload-registrar/controllers/pod_controller.go

```diff
@@ -0,0 +1,342 @@
+		Value: fmt.Sprintf("%s:%s", selector, value),
+	}
+}
+
+func (r *PodReconciler) selectorsToNamespacedName(selectors []*common.Selector) *types.NamespacedName {
+	podNamespace := ""
+	podName := ""
+	for _, selector := range selectors {
+		if selector.Type == "k8s" {
+			splitted := strings.SplitN(selector.Value, ":", 2)
+			if len(splitted) > 1 {
+				switch WorkloadSelectorSubType(splitted[0]) {
+				case PodNamespaceSelector:
+					podNamespace = splitted[1]
+				case PodNameSelector:
+					podName = splitted[1]
+				}
+			}
+		}
+	}
+	if podNamespace != "" && podName != "" {
+		return &types.NamespacedName{
+			Namespace: podNamespace,
+			Name:      podName,
+		}
+	}
+	return nil
+}
+
+func (r *PodReconciler) makeSpiffeID(obj ObjectWithMetadata) string {
+	return r.makeSpiffeIDForPod(obj.(*corev1.Pod))
+}
+
+func (r *PodReconciler) mungeIP(ip string) string {
+	if strings.Contains(ip, ".") {
+		// IPv4
+		return strings.Replace(ip, ".", "-", -1)
+	}
+	if strings.Contains(ip, ":") {
+		// IPv6
+		return strings.Replace(ip, ":", "-", -1)
+	}
+	return ip
+}
+
+func (r *PodReconciler) getNamesForEndpoints(ctx context.Context, pod *corev1.Pod) ([]string, error) {
+	names := make(map[string]bool)
+
+	endpointsList := corev1.EndpointsList{}
+	if err := r.List(ctx, &endpointsList, client.InNamespace(pod.Namespace), client.MatchingFields{endpointSubsetAddressReferenceField: pod.Name}); err != nil {
+		return nil, err
+	}
+
+	for _, endpoints := range endpointsList.Items {
+		endpoints := endpoints
+		// Based on https://github.com/kubernetes/dns/blob/master/docs/specification.md
+		// We cheat slightly and don't check the type of service (headless or not), we just add all possible names.
+
+		// 2.3.1 and 2.4.1: <service>.<ns>.svc.<zone>
+		names[fmt.Sprintf("%s.%s.svc.%s", endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+
+		r.forEachPodEndpointAddress(&endpoints, func(address corev1.EndpointAddress) {
+			if pod.Name == address.TargetRef.Name && pod.Namespace == address.TargetRef.Namespace {
+				// 2.4.1: <hostname>.<service>.<ns>.svc.<zone>
+				if address.Hostname != "" {
+					names[fmt.Sprintf("%s.%s.%s.svc.%s", address.Hostname, endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+				} else {
+					// The spec leaves this case up to the implementation, so here we copy CoreDns...
+					// CoreDNS has an endpoint_pod_names flag to switch between the following two options. We don't have that flag, so
+					// we'll just add both pod name and IP based name (the CoreDns default) for now.
+					names[fmt.Sprintf("%s.%s.%s.svc.%s", r.mungeIP(address.IP), endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+					names[fmt.Sprintf("%s.%s.%s.svc.%s", address.TargetRef.Name, endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+				}
+			}
+		})
+	}
+
+	namesSlice := make([]string, len(names))
+	i := 0
+	for name := range names {
+		namesSlice[i] = name
+		i++
+	}
+
+	return namesSlice, nil
```

Comment:
Is the order important here? even for debuggability should these entries be sorted?

Reply:
Elsewhere we do compare the dns names in an entry to establish if two entries are "identical enough", but we do it without checking order as I wasn't 100% sure I could rely on spire returning them in a stable order. I've added sorting here though, since you're right, it is cleaner. If we're sure the ordering will be maintained by spire (and the client) too then I'll alter the identity check in base_controller to be order dependent too.

Reply:
Because the 1st dnsname is added as the CN, SPIRE _should_ be maintaining the order (if not, its a bug). It's probably not a bad idea for this code to be robust in the face of that kind of bug though.